ramips: add support for Qihoo 360 P2

DragonBluep opened the pull request:
Specifications:
 * SOC: MT7628AN + MT7612EN
 * RAM: 128 MiB
 * Flash: Winbond W25Q128 (16MiB)
 * WLAN: 2*2 MIMO 2.4G (300M) + 2*2 MIMO 5G (867M)
 * LAN: four 10M/100M LAN ports
 * WAN: one 10M/100M WAN port
 * USB: 1* USB 2.0
 * Buttons: 2, Reset + Release USB (use as WPS in OpenWrt)
 * LEDs: 11
 * TTL: 1, Baudrate 57600
 * TFTP Server IP: 192.168.1.192

Installation:
1. Downgrade firmware to Ver 1.0.xx
2. Upgrade firmware to a special version Ver 1.0.42.42668, which can use
telnet. (search "360POP-P2-DEBUG-V1.0.42.42668.bin" to get it)
3. Use putty to open telnet (account: admin), upload a 3rd party bootloader
to /tmp by wget command and "hfs". (Due to factory bootloader would verify
firmware, and the firmware verify bootloader too, so a 3rd party bootloader
is needed. By the way, Qihoo 360 is a very famous computer security company,
I think it's ambitious to try to crack official bootloader. Recommended to
use "BREED" by HackPascal, search OpenWrt wiki. The bootloader compatible
with HIWIFI HC5661A)
/====================================================================
Notice! Before step 4!
a. If you want to backup official firmware, just use cat command to get mtd0
to /tmp, then use tftp command to transfer it to your computer
        cat /dev/mtd0 > /tmp/all.bin
        tftp -p -l /tmp/all.bin -r all.bin 192.168.x.x
b. Make sure everything is done otherwise your device may broken. It's
necessary to read back mtd1 to check if bootloader are write Correctly.
c. Do not try to use "dd" or "mtd write"!
/=====================================================================
4. Erase and write bootloader (Due to factory bootloader would verify firmware,
and the system verify bootloader too, so a 3rd party bootloader is
. By the way, Qihoo 360 is a very famous computer security company, it's
ambitious to try to crack official bootloader)
	mtd_write erase /dev/mtd1
	cat yourbootloader.bin > /dev/mtd1
5. Unplug the power, push reset button and plug in, hold for several seconds
to enter into bootloader's web recovery. if you do not flash 3rd party firmware,
the factory firmware will replace your bootloader to factory version when next
power on.
6. Flash openwrt*.bin, then it will boot.

adschm commented on target/linux/ramips/dts/mt7628an_360-p2.dts:
```diff
@@ -0,0 +1,138 @@
+
+/ {
+	compatible = "360,p2", "mediatek,mt7628an-soc";
+	model = "360 P2";
```
So, "360" is the vendor here?

DragonBluep replied:
Yes, the full name is "Qihoo 360 Technology Co. Ltd." but few of people call it "Qihoo" recent years.

adschm replied:
Looks like Qihoo is preferred, though: https://github.com/openwrt/openwrt/commit/a9360452f0130bb39a8e732334a174cc3f8991fd

DragonBluep replied:
okay, I changed it.

adschm commented on target/linux/ramips/dts/mt7628an_360-p2.dts:
```diff
@@ -0,0 +1,138 @@
+&state_default {
+	gpio {
+		ralink,group = "i2s", "refclk", "wdt", "wled_an", "uart1";
+		ralink,function = "gpio";
```
property names need to be updated.

DragonBluep replied:
OK, I fix it, it seems like changed from previous branch

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -29,6 +29,16 @@ define Build/ravpower-wd009-factory
+  DEVICE_MODEL := P2
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport \
+	kmod-mt76x2 
+  SUPPORTED_DEVICES += p2
```
This can be dropped.

DragonBluep replied:
So what's wrong here? I am confusing about it.

adschm replied:
Why do you add "p2" to the supported devices?

DragonBluep replied:
Has been removed. I saw many devices have this statement. Does it is not used to verify device model when apply sysupgrade?

adschm replied:
Yes, but do you know a device that uses "p2" as its compatible?

DragonBluep replied:
Finally I understand. The device define of ramips in makefile does not contain vendor name before v21.02. It will cause compatibility problem if not add SUPPORTED_DEVICES, but new support devices don't have this historical problem.
The commit has been force pushed and I think all issues have been solved.

adschm commented on target/linux/ramips/dts/mt7628an_360-p2.dts:
```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
```
File name is wrong, hyphen -> "_". Has this not been tested?

DragonBluep replied:
I got it, that's really a error here. I compiled OpenWrt v18.06 & v19.07 for this device some months ago and it can run well. I think it just a name error here. All gpio, port config and partition layout are totally right.

adschm commented:
Please remove the merge commits, so this only contains one proper commit.

musashino205 commented:
> Use putty to open telnet (account: admin), upload a 3rd party bootloader
to /tmp by wget command and "hfs". (recommended to use "BREED" by HackPascal,

Why does the original bootloader need to be replaced?

DragonBluep commented:
> > Use putty to open telnet (account: admin), upload a 3rd party bootloader
> > to /tmp by wget command and "hfs". (recommended to use "BREED" by HackPascal,
> 
> Why does the original bootloader need to be replaced?

Due to factory bootloader would verify firmware, and the firmware verify bootloader too, so a 3rd party bootloader is needed. By the way, Qihoo 360 is a very famous computer security company, I think it's ambitious to try to crack official bootloader. Recommended to use "BREED" by HackPascal, search OpenWrt wiki. The bootloader compatible with HIWIFI HC5661A.

DragonBluep commented:
Commit  has been updated  several days, need reviews. Yours Sincerely. @adschm @musashino205

DragonBluep commented:
Tested on master branch.

[    0.000000] Linux version 5.4.113 (dragon@debian) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r16485-cae3e37383)) #0 Mon Apr 19 12:46:37 2021
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7628AN ver:1 eco:2
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019655 (MIPS 24KEc)
[    0.000000] MIPS: machine is Qihoo P2
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000]   Normal zone: 288 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=00040000
[    0.000000] Readback ErrCtl register=00040000
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 122364K/131072K available (4741K kernel code, 198K rwdata, 1048K rodata, 1172K init, 197K bss, 8708K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] intc: using register map from devicetree
[    0.000000] random: get_random_bytes called from start_kernel+0x308/0x4fc with crng_init=0
[    0.000000] CPU Clock: 580MHz
[    0.000000] timer_probe: no matching timers found
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000010] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.015377] Calibrating delay loop... 385.02 BogoMIPS (lpj=770048)
[    0.059453] pid_max: default: 32768 minimum: 301
[    0.068786] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.083119] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.104903] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.124113] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.137621] pinctrl core: initialized pinctrl subsystem
[    0.149006] NET: Registered protocol family 16
[    0.267790] mt7620-pci 10140000.pcie: Port 0 N_FTS = 1b105000
[    0.428786] PCI host bridge /pcie@10140000 ranges:
[    0.438131]  MEM 0x0000000020000000..0x000000002fffffff
[    0.448441]   IO 0x0000000010160000..0x000000001016ffff
[    0.482495] PCI host bridge to bus 0000:00
[    0.490548] pci_bus 0000:00: root bus resource [mem 0x20000000-0x2fffffff]
[    0.504054] pci_bus 0000:00: root bus resource [io  0xffffffff]
[    0.515750] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.529163] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.544876] pci 0000:00:00.0: [14c3:0801] type 01 class 0x060400
[    0.556719] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x7fffffff]
[    0.569070] pci 0000:00:00.0: reg 0x14: [mem 0x20200000-0x2020ffff]
[    0.581531] pci 0000:00:00.0: supports D1
[    0.589367] pci 0000:00:00.0: PME# supported from D0 D1 D3hot
[    0.602474] pci 0000:01:00.0: [14c3:7662] type 00 class 0x028000
[    0.614326] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]
[    0.627686] pci 0000:01:00.0: reg 0x30: [mem 0x00000000-0x0000ffff pref]
[    0.641001] pci 0000:01:00.0: PME# supported from D0 D3hot D3cold
[    0.654477] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01
[    0.667469] pci_bus 0000:00: busn_res: [bus 00-ff] end is updated to 01
[    0.680538] pci 0000:00:00.0: BAR 0: no space for [mem size 0x80000000]
[    0.693575] pci 0000:00:00.0: BAR 0: failed to assign [mem size 0x80000000]
[    0.707337] pci 0000:00:00.0: BAR 8: assigned [mem 0x20000000-0x200fffff]
[    0.720753] pci 0000:00:00.0: BAR 9: assigned [mem 0x20100000-0x201fffff pref]
[    0.735028] pci 0000:00:00.0: BAR 1: assigned [mem 0x20200000-0x2020ffff]
[    0.748455] pci 0000:01:00.0: BAR 0: assigned [mem 0x20000000-0x200fffff 64bit]
[    0.762908] pci 0000:01:00.0: BAR 6: assigned [mem 0x20100000-0x2010ffff pref]
[    0.777166] pci 0000:00:00.0: PCI bridge to [bus 01]
[    0.786971] pci 0000:00:00.0:   bridge window [mem 0x20000000-0x200fffff]
[    0.800386] pci 0000:00:00.0:   bridge window [mem 0x20100000-0x201fffff pref]
[    0.814783] workqueue: max_active 576 requested for napi_workq is out of range, clamping between 1 and 512
[    0.837949] clocksource: Switched to clocksource MIPS
[    0.849198] NET: Registered protocol family 2
[    0.858690] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.875168] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.890244] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.904175] TCP: Hash tables configured (established 1024 bind 1024)
[    0.916876] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.929717] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.943771] NET: Registered protocol family 1
[    0.952376] PCI: CLS 0 bytes, default 32
[    0.964793] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.986126] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.997541] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    1.035071] mt7621_gpio 10000600.gpio: registering 32 gpios
[    1.046350] mt7621_gpio 10000600.gpio: registering 32 gpios
[    1.057612] mt7621_gpio 10000600.gpio: registering 32 gpios
[    1.068881] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    1.082656] printk: console [ttyS0] disabled
[    1.091091] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 28, base_baud = 2500000) is a 16550A
[    1.108896] printk: console [ttyS0] enabled
[    1.125372] printk: bootconsole [early0] disabled
[    1.145314] spi-mt7621 10000b00.spi: sys_freq: 193333333
[    1.172925] spi-nor spi0.0: w25q128 (16384 Kbytes)
[    1.182484] 4 fixed-partitions partitions found on MTD device spi0.0
[    1.195069] Creating 4 MTD partitions on "spi0.0":
[    1.204570] 0x000000000000-0x000000030000 : "bootloader"
[    1.216171] 0x000000030000-0x000000040000 : "config"
[    1.227082] 0x000000040000-0x000000050000 : "factory"
[    1.238302] 0x000000050000-0x000001000000 : "firmware"
[    1.252683] 2 uimage-fw partitions found on MTD device firmware
[    1.264488] Creating 2 MTD partitions on "firmware":
[    1.274338] 0x000000000000-0x0000001e7a1e : "kernel"
[    1.285245] 0x0000001e7a1e-0x000000fb0000 : "rootfs"
[    1.296101] mtd: device 5 (rootfs) set to be root filesystem
[    1.309084] 1 squashfs-split partitions found on MTD device rootfs
[    1.321418] 0x000000510000-0x000000fb0000 : "rootfs_data"
[    1.334068] libphy: Fixed MDIO Bus: probed
[    1.353725] rt3050-esw 10110000.esw: link changed 0x00
[    1.366041] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    1.384357] NET: Registered protocol family 10
[    1.397384] Segment Routing with IPv6
[    1.404858] NET: Registered protocol family 17
[    1.413736] 8021q: 802.1Q VLAN Support v1.8
[    1.429376] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    1.450160] Freeing unused kernel memory: 1172K
[    1.459157] This architecture does not have kernel memory protection.
[    1.471909] Run /sbin/init as init process
[    1.717962] random: fast init done
[    2.131295] init: Console is alive
[    2.138410] init: - watchdog -
[    3.165253] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.297377] usbcore: registered new interface driver usbfs
[    3.308441] usbcore: registered new interface driver hub
[    3.319094] usbcore: registered new device driver usb
[    3.336524] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    3.351467] ehci-fsl: Freescale EHCI Host controller driver
[    3.364304] ehci-platform: EHCI generic platform driver
[    3.385044] phy phy-10120000.usbphy.0: remote usb device wakeup disabled
[    3.398329] phy phy-10120000.usbphy.0: UTMI 16bit 30MHz
[    3.408681] ehci-platform 101c0000.ehci: EHCI Host Controller
[    3.420101] ehci-platform 101c0000.ehci: new USB bus registered, assigned bus number 1
[    3.435935] ehci-platform 101c0000.ehci: irq 26, io mem 0x101c0000
[    3.461974] ehci-platform 101c0000.ehci: USB 2.0 started, EHCI 1.00
[    3.475467] hub 1-0:1.0: USB hub found
[    3.483403] hub 1-0:1.0: 1 port detected
[    3.496913] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    3.511086] ohci-platform: OHCI generic platform driver
[    3.521792] ohci-platform 101c1000.ohci: Generic Platform OHCI controller
[    3.535315] ohci-platform 101c1000.ohci: new USB bus registered, assigned bus number 2
[    3.551142] ohci-platform 101c1000.ohci: irq 26, io mem 0x101c1000
[    3.626997] hub 2-0:1.0: USB hub found
[    3.634957] hub 2-0:1.0: 1 port detected
[    3.648060] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    3.672985] init: - preinit -
[    5.074936] random: jshn: uninitialized urandom read (4 bytes read)
[    5.254281] random: jshn: uninitialized urandom read (4 bytes read)
[    5.484929] random: jshn: uninitialized urandom read (4 bytes read)
[    5.886596] rt3050-esw 10110000.esw: link changed 0x00
[    9.816921] rt3050-esw 10110000.esw: link changed 0x02
[   10.218090] jffs2: notice: (443) jffs2_build_xattr_subsystem: complete building xattr subsystem, 6 of xdatum (0 unchecked, 1 orphan) and 7 of xref (1 dead, 0 orphan) found.
[   10.250259] mount_root: switching to jffs2 overlay
[   10.263604] overlayfs: upper fs does not support tmpfile.
[   10.279930] urandom-seed: Seeding with /etc/urandom.seed
[   10.430186] procd: - early -
[   10.436456] procd: - watchdog -
[   11.162061] procd: - watchdog -
[   11.169042] procd: - ubus -
[   11.253843] urandom_read: 3 callbacks suppressed
[   11.253852] random: ubusd: uninitialized urandom read (4 bytes read)
[   11.278378] random: ubusd: uninitialized urandom read (4 bytes read)
[   11.291612] random: ubusd: uninitialized urandom read (4 bytes read)
[   11.311273] procd: - init -
[   12.406197] kmodloader: loading kernel modules from /etc/modules.d/*
[   12.473560] Loading modules backported from Linux version v5.10.16-0-gde53befa79cf
[   12.488654] Backport generated by backports.git v5.10.16-1-0-g21d2a1d2
[   12.584573] xt_time: kernel timezone is -0000
[   12.825714] mt76_wmac 10300000.wmac: ASIC revision: 76280001
[   12.976813] urngd: v1.0.2 started.
[   13.186765] random: crng init done
[   13.193533] random: 1 urandom warning(s) missed due to ratelimiting
[   13.888364] mt76_wmac 10300000.wmac: Firmware Version: 20151201
[   13.900176] mt76_wmac 10300000.wmac: Build Time: 20151201183641
[   13.911910] mt76_wmac 10300000.wmac: Firmware already running...
[   13.923817] mt76_wmac 10300000.wmac: firmware init done
[   13.987226] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   14.036136] mt76x2e 0000:01:00.0: card - bus=0x1, slot = 0x0 irq=4
[   14.048746] mt76x2e 0000:01:00.0: ASIC revision: 76120044
[   15.262183] mt76x2e 0000:01:00.0: ROM patch build: 20141115060606a
[   15.295633] mt76x2e 0000:01:00.0: Firmware Version: 0.0.00
[   15.306575] mt76x2e 0000:01:00.0: Build: 1
[   15.314713] mt76x2e 0000:01:00.0: Build Time: 201507311614____
[   15.481987] mt76x2e 0000:01:00.0: Firmware running!
[   15.500163] ieee80211 phy1: Selected rate control algorithm 'minstrel_ht'
[   15.562842] PPP generic driver version 2.4.2
[   15.583079] NET: Registered protocol family 24
[   15.634695] kmodloader: done loading kernel modules from /etc/modules.d/*
[   22.598881] rt3050-esw 10110000.esw: link changed 0x00
[   26.531367] rt3050-esw 10110000.esw: link changed 0x02
[   29.774376] br-lan: port 1(eth0.1) entered blocking state
[   29.785142] br-lan: port 1(eth0.1) entered disabled state
[   29.796309] device eth0.1 entered promiscuous mode
[   29.805857] device eth0 entered promiscuous mode
[   29.942157] br-lan: port 1(eth0.1) entered blocking state
[   29.952907] br-lan: port 1(eth0.1) entered forwarding state
[   30.762641] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   36.361123] br-lan: port 2(wlan1) entered blocking state
[   36.371774] br-lan: port 2(wlan1) entered disabled state
[   36.382752] device wlan1 entered promiscuous mode
[   36.559636] br-lan: port 3(wlan0) entered blocking state
[   36.570245] br-lan: port 3(wlan0) entered disabled state
[   36.581240] device wlan0 entered promiscuous mode
[   36.590749] br-lan: port 3(wlan0) entered blocking state
[   36.601308] br-lan: port 3(wlan0) entered forwarding state
[   36.694862] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   37.444770] IPv6: ADDRCONF(NETDEV_CHANGE): wlan1: link becomes ready
[   37.457687] br-lan: port 2(wlan1) entered blocking state
[   37.468262] br-lan: port 2(wlan1) entered forwarding state

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -325,6 +325,15 @@ define Device/onion_omega2p
+  DEVICE_VENDOR := Qihoo
+  DEVICE_MODEL := P2
+  DEVICE_PACKAGES := kmod-mt76x2 kmod-usb-ohci kmod-usb2 \
+  kmod-usb-ledtrig-usbport
```
Single tab for hanging indent.

DragonBluep replied:
I will try to make it work in the future, but maybe I need to suspend this PR for a while before that.

adschm commented:
Replacing the bootloader typically is a no-go on this target.

DragonBluep commented:
> Replacing the bootloader typically is a no-go on this target.

I will try to make it work in the future, but maybe I need to suspend this PR for a while before that.

JackieCat0418 commented:
wow，nice

DragonBluep committed:
ramips: add support for Qihoo 360 P2

Specifications:
 * SOC: MT7628AN + MT7612EN
 * RAM: 128 MiB
 * Flash: Winbond W25Q128 (16MiB)
 * WLAN: 2*2 MIMO 2.4G (300M) + 2*2 MiMO 5G (867M)
 * LAN: four 10M/100M LAN ports
 * WAN: one 10M/100M WAN port
 * USB: 1* USB 2.0
 * Buttons: 2, Reset + Release USB (use as WPS in OpenWrt)
 * LEDs: 11
 * TTL: 1, baudrate 57600
 * TFTP Server IP: 192.168.1.192

Installation:
1. Downgrade firmware to Ver 1.0.xx
2. Upgrade firmware to a special version Ver 1.0.42.42668, which can use
telnet. (search "360POP-P2-DEBUG-V1.0.42.42668.bin" to get it)
3. Use putty to open telnet (account: admin), upload a 3rd party bootloader
to /tmp by wget command and "hfs". (Due to factory bootloader would verify
firmware, and the firmware verify bootloader too, so a 3rd party bootloader
is needed. By the way, Qihoo 360 is a very famous computer security company,
I think it's ambitious to try to crack official bootloader. Recommended to
use "BREED" by HackPascal, search OpenWrt wiki. The bootloader compatible
with HIWIFI HC5661A)
/====================================================================
Notice! Before step 4!
a. If you want to backup official firmware, just use cat command to get mtd0
to /tmp, then use tftp command to transfer it to your computer
        cat /dev/mtd0 > /tmp/all.bin
        tftp -p -l /tmp/all.bin -r all.bin 192.168.x.x
b. Make sure everything is done otherwise your device may broken. It's
necessary to read back mtd1 to check if bootloader are write Correctly.
c. Do not try to use "dd" or "mtd write"!
/=====================================================================
4. Erase and write bootloader (Due to factory bootloader would verify firmware,
and the system verify bootloader too, so a 3rd party bootloader is
. By the way, Qihoo 360 is a very famous computer security company, it's
ambitious to try to crack official bootloader)
	mtd_write erase /dev/mtd1
	cat yourbootloader.bin > /dev/mtd1
5. Unplug the power, push reset button and plug in, hold for several seconds
to enter into bootloader's web recovery. if you do not flash 3rd party firmware,
the factory firmware will replace your bootloader to factory version when next
power on.
6. Flash openwrt*.bin, then it will boot.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
Close for a while, I am busy now and need a long time to adapt official bootloader.
